Replaces pint with astropy.units

pkj-m opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/radis/blob/develop/CONTRIBUTING.md . -->

### Description
<!-- Provide a general description of what your pull request does. -->

- This PR replaces `pint` with `astropy.units` in the `phys/units.py` file.
- The import has been removed from the file and replaced with astropy units everywhere. 
- The `DimensionalityError` offered by `pint` has been replaced with astropy's `UnitConversionError` everywhere.

A few tests were breaking as they depended on pint's `UnitRegistry` which was imported as `Q_` earlier to convert the strings to units. The same task (and name) has now been given to astropy's `units.Unit` class which acts as the equivalent.

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

### TODO: 
- [x]  clean the code and remove old comments after review
- [x] search for other methods using `pint` and replace them with `astropy.units`
- [x] remove `pint` from dependencies

Will fix #104

erwanp commented on radis/test/test_phys.py:
```diff
@@ -116,8 +117,9 @@ def test_units(verbose=True, *args, **kwargs):
     # RADIS pint-aware array
-    a = uarray(np.linspace(10, 100, 10), "Td")
-    res = Q_(np.linspace(1e-16, 1e-15, 10), "V * cm^2")  # pint definition
+    Td = "1e-17 * volt * centimeter**2"
+    a = uarray(np.linspace(10, 100, 10), Td)
+    res = np.linspace(1e-16, 1e-15, 10) * u.Unit("V * cm^2")  # pint definition
 
     assert (
         np.round(np.array(a.to("V * cm^2")) - np.array(res), 5) == np.zeros_like(res)
```
```suggestion
```

these can be removed, they're more plasma physics than spectroscopy.

erwanp commented on radis/phys/units.py:
```diff
@@ -114,34 +106,29 @@ def conv2(quantity, fromunit, tounit):
+        a = quantity * u.Unit(fromunit)
+        a = a.to(u.Unit(tounit))
 
         # Hardcoded: 'pint' considers angles to be dimensionless (which they are)
```
I think unlike pint, astropy handles Angles as a physical unit by default. So the comment could be removed. 

Add to the test something like 
```
assert mW/cm2/sr/nm not equivalent to mW/cm2/nm  ?  # pseudo-code
```

erwanp commented:
I added some changes above. 

Now I have a big remark:

### Non-standard units

Whatever Astropy.units may consider standard, we do not want to have to write
```
mW/cm**2/sr/nm
```
That's just very unreadable compared to
```
mW/cm2/sr/nm
```
There are a few like this:
```
cm**2 -> cm2
m**2 -> m2
cm**3 -> cm3
m**3 -> m3 
mm**3 -> mm3 
```

And of course the famous "cm-1" cherished by all infrared spectroscopists. 
`cm-1 -> cm**-1 `


Preventing the use of this syntax would certainly cut us from all our user base (@minouHub @gugzy first ). 


So the question is more: 

**How do we implement these in an _elegant_, and _safe_ way?**

- Astropy allows custom units. But setting `m3` as `m**3` may not be safe: `cm3` would be interpreted as 1/100 * m**3 and not (1/100*m)**3 
- Currently, were were parsing & replacing text. That may still work. They'll only be given as inputs. We want:
```
substract(s, 2, "mW/cm2/sr/nm")   # fast version
substract(s, 2* u.mW/u.cm**2/u.sr/u.nm)  # pure astropy version
```
but we don't really need
```
substract(s, 2* u.mW/u.cm2/u.sr/u.nm)  # if really you added all these u. you can afford u.cm**2 
``` 

Same for the "cm-1": 
```
substract(s, 2, "mW/cm2/sr/cm-1")   # would still work 
substract(s, 2* u.mW/u.cm**2/u.sr/(1/u.cm))  # astropy units version 
```
but good luck in parsing the next:
```
substract(s, 2* u.mW/u.cm**2/u.sr/u.cm-1)
``` 

Any better idea than parsing text?

pkj-m commented:
hey @erwanp , thanks for your detailed remarks. However, I believe that astropy does infact support units the way you would like to see in Radis.

So if we choose to handle the non-standard units in a format such as `mW/cm2/sr/nm`, it's definitely possible, as can be quickly checked with something like `10 * u.Unit("mW/cm2/sr/nm")` which returns us exactly what we would expect: 
![image](https://user-images.githubusercontent.com/17577971/79509109-80868b80-8058-11ea-8a26-5d8d6bc2a42a.png)

This also includes writing centimeter inverse as `"cm-1"`.

Also, could you explain your point regarding the possibility of confusing `cm3` as `(1/100)*m^3` instead of `(1/100 * m)^3`. Is that directed at users who might get confused? Regardless, I think Astropy does a good job at handling these unit formats as we can see:
![image](https://user-images.githubusercontent.com/17577971/79510328-cf352500-805a-11ea-9bfa-68a1058387be.png)
and 
![image](https://user-images.githubusercontent.com/17577971/79510574-3a7ef700-805b-11ea-9022-77abe24d2824.png)



The reason I changed the syntax from `cm3` to `cm**3` was because I felt it looked more mathematically accurate. But you're right in pointing out that it makes the function calls significantly uglier for users.

Does that resolve the problem or did I misinterpret something?

pkj-m commented:
Another thing that I found while going through the CI tests output was the use of `cm_1` in a lot of places in the code instead for `cm-1`. I think `pint` had a way of parsing `cm_1` as `cm-1` from the `Units.txt` file that was in the `phys` folder. However, I couldn't come up with a quicker and cleaner method to solve the problem than to simply replace all occurences of `cm_1` in the code base to `cm-1`. Is that alright or do you think we should handle this problem in a different manner?

erwanp commented:
> 
> Also, could you explain your point regarding the possibility of confusing `cm3` as `(1/100)*m^3` instead of `(1/100 * m)^3`. Is that directed at users who might get confused? 

I was concerned that by defining a unit such as m3 ~ m**3 the parser would ready cm3 as c(m3) instead of (cm)3. But as you found out, it turns out I was worried for no reason: 

> Regardless, I think Astropy does a good job at handling these unit formats

Replacing "cm_1" with "cm-1" is actually better than initially. "cm_1" was there because `pint` couldnt resolve `cm-1`. Astropy can, so let's keep it! 

Nice improvement!

erwanp commented:
The test crash on
```assert s.get_power(unit="W/cm2/sr") == 2631.6288408588148```
You should replace it with `np.isclose()` and that will do.

pkj-m commented:
Hey @erwanp , I was trying to fix the spectrum tests that are failing. It seems that the `apply_slit` method calls `pint`'s `UnitRegistry.units` method which basically returns the unit in a string format which was used further ([exact line here](https://github.com/radis/radis/blob/develop/radis/spectrum/spectrum.py#L2400)).

After changing that line to do the exact same thing except with `astropy` units this time, I ran into another error:
```astropy.units.core.UnitConversionError: 'mW / (cm2 nm sr um)' and 'mW / (cm2 sr)' are not convertible``` thrown by the `conv2` method.

At some point, the test is trying to convert a quantity from one unit to another one with different dimensions.  On printing the exact unit of the quantity which is being handled in the `apply_slit` function, I found out it's `mW/cm2/sr/µm*nm`. However, I am not sure why this was being converted to `mW/cm2/sr`. 
I think I am confused about the logic behind the transformations, and even though you did explain about it above, I think I am still not clear on how to handle this particular problem.

erwanp commented:
So that's what we're doing when applying the slit:
![image](https://user-images.githubusercontent.com/16088743/79614756-7ff5ff80-8101-11ea-97ca-b44cd5a07765.png)
In terms of units, the spectrum after slit as the [initial unit] * [unit of convolution kernel Phi] * [unit of wavespace]

Now there are two ways to normalize the convolution kernel. Either by area: 

![image](https://user-images.githubusercontent.com/16088743/79614801-9dc36480-8101-11ea-9958-a57f55fc0394.png)

Either at maximum: 

![image](https://user-images.githubusercontent.com/16088743/79614822-a87df980-8101-11ea-8bfd-66cf66b797c1.png)

- So in the first case we end up with no change in inputs. 
- In the second case we need to multiply the spectrum unit by [unit of wavespace]   (can be `nm`, `cm-1`, or whatever was used). 

This is what we're doing. 

---

What surprises me in your error message is we end up with `'mW / (cm2 nm sr um)' . While we should have had 'mW / (cm2 sr um) * nm'    (which is indeed homogeneous to  'mW / (cm2 sr)')

pkj-m commented:
Hi @erwanp ,  thanks for the thorough explanation. I understand it much better now!

> What surprises me in your error message is we end up with 'mW / (cm2 nm sr um) . While we should have had mW / (cm2 sr um) * nm (which is indeed homogeneous to mW / (cm2 sr))

That problem got fixed by rearranging the units such so that it read `nm * mW / cm2 / sr / um`. 
However, I have run into some other problems now. This is the console output when I run the `test_slit.py` file locally after substituting all occurrences of `I/I0`, `-ln(I/I0)` and `eps` from the defaults' dictionary in the BaseFactory class, and also while loading them from the database loader:

```
...
>>> _test_all_slits yield correct FWHM (+- wstep) : OK

File CO_Tgas1500K_mole_fraction0.01.spec has a deprecrated structure (key medium removed in 0.9.22). Fixing this time, but regenerate database ASAP.
File CO_Tgas1500K_mole_fraction0.01.spec has a deprecrated structure (v1u in lines is now called vu). Fixed this time, but regenerate database ASAP.
/home/pankaj/radis/radis/spectrum/rescale.py:1967: UserWarning: Large rescaling from 0.01 to 1. There may be significant changes in pressure broadening coefficients.You should calculate a new spectrum instead.
  + "You should calculate a new spectrum instead."
WARNING: UnitsWarning: 'mW/cm2/sr/nm' contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
WARNING: UnitsWarning: 'mW/sr/cm2/nm' contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
CO_Tgas1500K_mole_fraction0.01.spec new quantities added: ['absorbance', 'transmittance_noslit', 'emissivity_noslit', 'radiance_noslit', 'emisscoeff']
compare radiance with rtol=0.001
... couldn't plot radiance
compare radiance with rtol=0.001
... couldn't plot radiance
compare radiance with rtol=0.001
/home/pankaj/radis/radis/spectrum/compare.py:861: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).
  fig = plt.figure(num=nfig, figsize=figsize)
compare radiance with rtol=0.001
/home/pankaj/radis/radis/spectrum/compare.py:861: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).
  fig = plt.figure(num=nfig, figsize=figsize)
File CO_Tgas1500K_mole_fraction0.5.spec has a deprecrated structure (key medium removed in 0.9.22). Fixing this time, but regenerate database ASAP.
File CO_Tgas1500K_mole_fraction0.5.spec has a deprecrated structure (v1u in lines is now called vu). Fixed this time, but regenerate database ASAP.
WARNING: UnitsWarning: 'mW/cm2/sr/nm' contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
WARNING: UnitsWarning: 'mW/sr/cm2/nm' contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
WARNING: UnitsWarning: 'nm*mW/cm2/sr/nm' contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
WARNING: UnitsWarning: 'mW/cm2/sr' contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
Test auto resampling
Using database: HITRAN-CO-TEST
HITRAN-CO-TEST 
-------
info : HITRAN 2016 database, CO, 3 main isotopes (CO-26, 36, 28), 2000-2300 cm-1
path : ['/home/pankaj/radis/radis/test/files/hitran_co_3iso_2000_2300cm.par']
format : hitran
parfuncfmt : hapi
levelsfmt : radis


Using cache file: /home/pankaj/radis/radis/test/files/hitran_co_3iso_2000_2300cm.h5
Calculating Equilibrium Spectrum
Physical Conditions
----------------------------------------
   Tgas                 300 K
   Trot                 300 K
   Tvib                 300 K
   isotope              1,2,3
   mole_fraction        0.02
   molecule             CO
   path_length          100 cm
   pressure_mbar        1013.25 mbar
   rot_distribution     boltzmann
   self_absorption      True
   state                X
   vib_distribution     boltzmann
   wavenum_max          2260.0000 cm-1
   wavenum_min          2230.0000 cm-1
Computation Parameters
----------------------------------------
   Tref                 296 K
   broadening_max_width  20 cm-1
   cutoff               1e-27 cm-1/(#.cm-2)
   db_assumed_sorted    True
   dbformat             hitran
   dbpath               /home/pankaj/radis/radis/test/files/hitran_co_3iso_2000_2300cm.par
   dlm_res_G            0.01
   dlm_res_L            0.01
   include_neighbouring_lines  True
   levelsfmt            radis
   parfuncfmt           hapi
   pseudo_continuum_threshold  0
   wavenum_max_calc     2270.0000 cm-1
   wavenum_min_calc     2220.0000 cm-1
   waveunit             cm-1
   wstep                0.02 cm-1
Information
----------------------------------------
   db_use_cached        True
   export_lines         True
   load_energies        True
   lvl_use_cached       True
   parallel             False
   warning_broadening_threshold  0.01
   warning_linestrength_cutoff  0.01
----------------------------------------
WARNING: UnitsWarning: 'mW/cm2/sr/nm' contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
WARNING: UnitsWarning: 'mW/sr/cm2/nm' contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
Spectrum calculated in 0.03s
/home/pankaj/radis/radis/spectrum/spectrum.py:3493: UserWarning: Wavespace is not evenly spaced (0.001%) for transmittance_noslit. This may create problems when convolving with slit function
  + " This may create problems when convolving with slit function"
test_slit.py:705: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).
  fig = plt.figure(fig_prefix + "auto-resampling")
Traceback (most recent call last):
  File "/home/pankaj/anaconda3/envs/radis-dev/lib/python3.6/site-packages/pyparsing.py", line 1683, in _parseNoCache
    loc, tokens = self.parseImpl(instring, preloc, doActions)
  File "/home/pankaj/anaconda3/envs/radis-dev/lib/python3.6/site-packages/pyparsing.py", line 2897, in parseImpl
    if instring[loc] == self.firstMatchChar:
IndexError: string index out of range

During handling of the above exception, another exception occurred:

pyparsing.ParseException: Expected {snowflake | accent | symbol | c_over_c | function | group | frac | dfrac | stackrel | binom | genfrac | sqrt | overline | operatorname}, found end of text  (at char 18), (line:1, col:19)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/pankaj/anaconda3/envs/radis-dev/lib/python3.6/site-packages/matplotlib/mathtext.py", line 2581, in parse
    result = self._expression.parseString(s)
  File "/home/pankaj/anaconda3/envs/radis-dev/lib/python3.6/site-packages/pyparsing.py", line 1955, in parseString
    raise exc
  File "/home/pankaj/anaconda3/envs/radis-dev/lib/python3.6/site-packages/pyparsing.py", line 4065, in parseImpl
    raise ParseSyntaxException._from_exception(pe)
pyparsing.ParseSyntaxException: Expected {snowflake | accent | symbol | c_over_c | function | group | frac | dfrac | stackrel | binom | genfrac | sqrt | overline | operatorname}, found end of text  (at char 18), (line:1, col:19)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test_slit.py", line 824, in <module>
    print(("Testing slit.py: ", _run_testcases(plot=True)))
  File "test_slit.py", line 818, in _run_testcases
    test_resampling(plot=plot, verbose=verbose, *args, **kwargs)
  File "test_slit.py", line 714, in test_resampling
    label="(stored in cm-1)",
  File "/home/pankaj/radis/radis/spectrum/spectrum.py", line 1654, in plot
    fix_style(str("origin"))
  File "/home/pankaj/anaconda3/envs/radis-dev/lib/python3.6/site-packages/publib/main.py", line 195, in fix_style
    _fix_style(style, ax, **kwargs)
  File "/home/pankaj/anaconda3/envs/radis-dev/lib/python3.6/site-packages/publib/main.py", line 230, in _fix_style
    plt.tight_layout()
  File "/home/pankaj/anaconda3/envs/radis-dev/lib/python3.6/site-packages/matplotlib/pyplot.py", line 1337, in tight_layout
    gcf().tight_layout(pad=pad, h_pad=h_pad, w_pad=w_pad, rect=rect)
  File "/home/pankaj/anaconda3/envs/radis-dev/lib/python3.6/site-packages/matplotlib/cbook/deprecation.py", line 358, in wrapper
    return func(*args, **kwargs)
  File "/home/pankaj/anaconda3/envs/radis-dev/lib/python3.6/site-packages/matplotlib/figure.py", line 2493, in tight_layout
    pad=pad, h_pad=h_pad, w_pad=w_pad, rect=rect)
  File "/home/pankaj/anaconda3/envs/radis-dev/lib/python3.6/site-packages/matplotlib/tight_layout.py", line 360, in get_tight_layout_figure
    pad=pad, h_pad=h_pad, w_pad=w_pad)
  File "/home/pankaj/anaconda3/envs/radis-dev/lib/python3.6/site-packages/matplotlib/tight_layout.py", line 109, in auto_adjust_subplotpars
    tight_bbox_raw = union([ax.get_tightbbox(renderer) for ax in subplots
  File "/home/pankaj/anaconda3/envs/radis-dev/lib/python3.6/site-packages/matplotlib/tight_layout.py", line 110, in <listcomp>
    if ax.get_visible()])
  File "/home/pankaj/anaconda3/envs/radis-dev/lib/python3.6/site-packages/matplotlib/axes/_base.py", line 4317, in get_tightbbox
    bb_xaxis = self.xaxis.get_tightbbox(renderer)
  File "/home/pankaj/anaconda3/envs/radis-dev/lib/python3.6/site-packages/matplotlib/axis.py", line 1202, in get_tightbbox
    *ticklabelBoxes2,
  File "/home/pankaj/anaconda3/envs/radis-dev/lib/python3.6/site-packages/matplotlib/axis.py", line 1200, in <genexpr>
    if a.get_visible()),
  File "/home/pankaj/anaconda3/envs/radis-dev/lib/python3.6/site-packages/matplotlib/text.py", line 905, in get_window_extent
    bbox, info, descent = self._get_layout(self._renderer)
  File "/home/pankaj/anaconda3/envs/radis-dev/lib/python3.6/site-packages/matplotlib/text.py", line 300, in _get_layout
    clean_line, self._fontproperties, ismath=ismath)
  File "/home/pankaj/anaconda3/envs/radis-dev/lib/python3.6/site-packages/matplotlib/backends/backend_agg.py", line 209, in get_text_width_height_descent
    self.mathtext_parser.parse(s, self.dpi, prop)
  File "/home/pankaj/anaconda3/envs/radis-dev/lib/python3.6/site-packages/matplotlib/mathtext.py", line 3354, in parse
    box = self._parser.parse(s, font_output, fontsize, dpi)
  File "/home/pankaj/anaconda3/envs/radis-dev/lib/python3.6/site-packages/matplotlib/mathtext.py", line 2586, in parse
    str(err)]))
ValueError: 
^\mathregular{-I/I
                  ^
Expected {snowflake | accent | symbol | c_over_c | function | group | frac | dfrac | stackrel | binom | genfrac | sqrt | overline | operatorname}, found end of text  (at char 18), (line:1, col:19)

```

Apologies for the extremely lengthy output and traceback. Any ideas why the tests are failing at this point? I can push the local changes to the remote branch if you'd like to go through them as well. 
Thanks!

erwanp commented:
The problem happens with the test file: `CO_Tgas1500K_mole_fraction0.01.spec`. 
Can you try to load it and plot it directly?
```
from radis.test.utils import getTestFile
from radis import load_spec
load_spec(`CO_Tgas1500K_mole_fraction0.01.spec`).plot()
```

Anyway, you should still push and I'll have a look!

pkj-m commented:
@erwanp you were right. I tried loading the spectrum directly and it gave the same error. Most probably has something to do with the changes I made while substituting the non-standard units with `1`. Also I have pushed the commits I made while debugging but it seems the merge with the updated develop branch has made it very cluttered. I think I'll clean up the history once I get the tests running or probably create another PR and close this one!

erwanp commented:
Ok I'll look that up later today, thanks for pushing!

erwanp commented:
Couple things: 

- it seems that astropy (unlike pint) does not recognizes "µm": we should write "um" instead. Note that there is a [make_up](https://github.com/radis/radis/blob/develop/radis/spectrum/utils.py#L145) function applied to the spectrum plots, where we could replace the "um" in "µm" at plotting time. 

- The error you saw above is created in this same make_up function. In particular I end up with:

```
'Wavenumber (cm$^\\mathregular{-I/I$_\\mathregular{0}$}$)'
```
I think it's because of the automating renaming of L161: label.replace(r"1", r"I/I$_\mathregular{0}$")

I could suggest that you re-introduce the "I/I0" as a unit, at plotting time, in the plot() function (around L1591):
```
ylabel = make_up("{0} ({1})".format(var, Iunit))
```
Maybe with:
```
if var in ["transmittance", "transmittance_noslit"] and unit=='1':
    Iunit = "I/I0"    # more explicit for the user
```

pkj-m commented:
@erwanp finally got the whole thing working! Thanks for all the help! Learnt quite a few new things about the package with this issue.
 Also, do you think I should open another PR with a cleaner git history given this PR got a little messed up due to all the merging with develop branch?

erwanp commented:
I think you could get rid of that with a proper rebase. Have a look at the [Einsteinpy Refactor Commits page](https://github.com/einsteinpy/einsteinpy/wiki/How-to-refactor-commits) (I referenced it on our wiki)  [ what they call `master` is our `develop`, though! ]

erwanp commented on radis/phys/units.py:
```diff
@@ -207,35 +194,36 @@ def convert_emi2cm(j_nm, wavenum, Iunit0, Iunit):
 
     # Convert to whatever was wanted
-    j_cm = conv2(j_cm, "mW/cm3/sr/cm_1", Iunit)
+    j_cm = conv2(j_cm, "mW/cm**3/sr/cm-1", Iunit)
```
```suggestion
    j_cm = conv2(j_cm, "mW/cm3/sr/cm-1", Iunit)
```

erwanp commented on radis/phys/units.py:
```diff
@@ -269,18 +257,19 @@ def convert_emi2nm(j_cm, wavenum, Iunit0, Iunit):
 
     # Convert to whatever was wanted
-    j_nm = conv2(j_nm, "mW/cm3/sr/nm", Iunit)
+    j_nm = conv2(j_nm, "mW/cm**3/sr/nm", Iunit)
```
```suggestion
    j_nm = conv2(j_nm, "mW/cm3/sr/nm", Iunit)
```

erwanp commented on radis/phys/units.py:
```diff
@@ -390,20 +380,21 @@ def convert_rad2nm(l_cm, wavenum, Iunit0, Iunit):
 
     # Convert to whatever was wanted
-    l_nm = conv2(l_nm, "mW/cm2/sr/nm", Iunit)
+    l_nm = conv2(l_nm, "mW/cm**2/sr/nm", Iunit)
```
```suggestion
    l_nm = conv2(l_nm, "mW/cm2/sr/nm", Iunit)
```

erwanp commented on radis/phys/units.py:
```diff
@@ -413,8 +404,8 @@ def convert_universal(
     spec=None,
-    per_nm_is_like="mW/sr/cm2/nm",
-    per_cm_is_like="mW/sr/cm2/cm_1",
+    per_nm_is_like="mW/sr/cm**2/nm",
+    per_cm_is_like="mW/sr/cm**2/cm-1",
```
```suggestion
    per_nm_is_like="mW/sr/cm2/nm",
    per_cm_is_like="mW/sr/cm2/cm-1",
```

erwanp commented on radis/spectrum/compare.py:
```diff
@@ -979,8 +987,9 @@ def get_wdiff_Idiff():
     if title:
         fig.suptitle(title)
-
+    # print("CALLED PLOT DIFF FUNCTION///...........")
```
```suggestion
```

erwanp commented on radis/spectrum/compare.py:
```diff
@@ -903,6 +908,9 @@ def get_wdiff_Idiff():
 
+    if var in ["transmittance", "transmittance_noslit"] and wunit == "1":
+        Iunit = "I/I0"  # more explicit for the user
+
```
```suggestion
    elif var == 'abscoeff' and wunits == "1":
        Iunit = "-ln(I/I0)" # more explicit for the user
    elif var in ["emissivity_no_slit", "emissivity"] and wunits == "1":
        Iunit = "eps"  # more explicit for the user

```

erwanp commented on radis/tools/database.py:
```diff
@@ -528,7 +535,7 @@ def _json_to_spec(sload, file=""):
 
     conditions = sload["conditions"]
-
+    # print("TILL LINE 535")
```
```suggestion
```

erwanp commented on radis/tools/database.py:
```diff
@@ -464,6 +464,7 @@ def load_spec(file, binary=False):  # , return_binary_status=False):
 
 
     """
+    # print("LOAD SPEC CALLED")
```
```suggestion
```

erwanp commented on radis/tools/database.py:
```diff
@@ -500,8 +501,14 @@ def _load(binary):
     sload, fixed = _fix_format(file, sload)
 
     # Generate Spectrum
+    # print("CALLING jsontospec function")
```
```suggestion
```

erwanp commented on radis/tools/database.py:
```diff
@@ -500,8 +501,14 @@ def _load(binary):
+    for unit in s.units:
+        if s.units[unit] in ["I/I0", "-ln(I/I0)", "eps"]:
+            s.units[unit] = "1"
+    # print("INSIDE THE LOAD FUNCTION, s.units = ", s.units)
```
```suggestion
```

erwanp commented on radis/tools/database.py:
```diff
@@ -500,8 +501,14 @@ def _load(binary):
+    # print("CALLING jsontospec function")
     s = _json_to_spec(sload, file)
 
+    # print("BACK FROM jsontospec")
```
```suggestion
```

erwanp commented on radis/tools/database.py:
```diff
@@ -559,7 +566,7 @@ def _json_to_spec(sload, file=""):
     # Generate spectrum:
     waveunit = sload["conditions"]["waveunit"]
-
+    # print("WAVEUNIT == ", waveunit)
```
```suggestion
```

erwanp commented on radis/tools/database.py:
```diff
@@ -570,13 +577,16 @@ def _json_to_spec(sload, file=""):
     else:
         slit = {}
 
+    # print("TILL 579...")
+
```
```suggestion
```

erwanp commented on radis/tools/database.py:
```diff
@@ -570,13 +577,16 @@ def _json_to_spec(sload, file=""):
     else:
         kwargs["lines"] = None
 
+    # print("TILL 588...")
```
```suggestion
```

erwanp commented on radis/tools/database.py:
```diff
@@ -595,20 +605,24 @@ def _json_to_spec(sload, file=""):
     else:
         kwargs["populations"] = None
 
+    # print("TILL 607...")
```
```suggestion
```

erwanp commented on radis/tools/database.py:
```diff
@@ -595,20 +605,24 @@ def _json_to_spec(sload, file=""):
         except KeyError:
             kwargs[attr] = None
 
+    # print("TILL 615...")
```
```suggestion
```

erwanp commented on radis/tools/database.py:
```diff
@@ -595,20 +605,24 @@ def _json_to_spec(sload, file=""):
         quantities=quantities, conditions=conditions, waveunit=waveunit, **kwargs
     )
 
+    # print("TILL 620...")
```
```suggestion
```

erwanp commented on radis/tools/database.py:
```diff
@@ -595,20 +605,24 @@ def _json_to_spec(sload, file=""):
     # ... add file
     s.file = basename(file)
 
+    # print("TILL 624...")
```
```suggestion
```

pkj-m commented on radis/spectrum/spectrum.py:
```diff
@@ -1597,7 +1603,7 @@ def plot(
             Iunit = "norm"
 
-        set_style("origin")
+        # set_style("origin")
```
remove #

pkj-m commented on radis/test/spectrum/test_spectrum.py:
```diff
@@ -402,17 +402,18 @@ def _run_testcases(
     # -------------
     test_intensity_conversion(debug=debug, verbose=verbose, *args, **kwargs)
-
+    print("LINE 405 done...")
```
remove comments from file

pkj-m commented on radis/spectrum/utils.py:
```diff
@@ -15,7 +15,7 @@
 # Waverange, wavespaces
 # ... aliases:
-WAVENUM_UNITS = ["cm", "cm-1", "cm_1", "wavenumber"]
+WAVENUM_UNITS = ["cm", "cm-1", "cm-1", "wavenumber"]
```
remove repetition

erwanp commented on radis/spectrum/spectrum.py:
```diff
@@ -2404,15 +2404,20 @@ def apply_slit(
-                new_unit = "{0}*{1}".format(
-                    self.units[qns], unit.replace("cm-1", "cm-1")
-                )
+                if self.units[qns] == "1":
+                    new_unit = "{0}*{1}".format(
+                        self.units[qns], unit.replace("cm-1", "cm-1")
+                    )
+                else:
+                    new_unit = "{0}*{1}".format(
+                        unit.replace("cm-1", "cm-1"), self.units[qns]
+                    )
```
Comments on the code:

- you should look up these `replace("cm-1", "cm-1")`  (they come from "cm_1" > "cm-1" ) and remove them. 

- do we really need the if/else here? 

---

There is also an underlying physical question, which requires to understand how experiments are done [ this will become a different Issue ]  : 

In my [comment](https://github.com/radis/radis/pull/106#issuecomment-615469626) I explained why the unit changes when you apply an instrumental slit function (to account for the spectrometer limited spectral resolution), if this slit function was normalized at the maximum instead of per area [ note that 99% of the people normalize at the area, we're dealing with a very particular case here, used only when calculating emission spectra to have an idea of the spectral flux]. This applied well to emission spectra.

Here, units == "1" refers to a transmittance spectra (or emissivity, but a similar reasoning applies): 

A theoretical transmittance spectrum is the ratio of an intensity I and a background I0. In practice, no-one ever measures a transmission spectrum directly. Your instrument collects an intensity I first, a background intensity I0 next, divides both and returns a transmittance spectrum I/I0. So your spectral resolution loss applies to I and I0. Your real transmittance spectrum is : `Filter(I) / Filter(I0)` instead of `Filter(I/I0)` as we are doing here. 
In practice the background is quite flat, so we have `Filter(I0)` ~ `I0` , so we end up with `Filter(I)/Filter(I0) ~ Filter(I)/I0 = Filter(I/I0)` , so applying the spectral filter to the theoretical transmittance spectrum directly is eventually a very good approximation.  But notice that  `Filter(I0)` ~ `I0`  assumed the filtering did not change the units! 

Therefore: 

- applying a slit not normalized by area to a transmittance spectrum is meaningless.  (people experienced in absorption spectroscopy like @dcmvdbekerom  will probably confirm!). 

- we want to remain numerically correct no matter what, so we should keep on correcting the unit as implemented here.  

- but we'll end up with transmittance spectra in "nm" or "cm-1" and with values that are not within [0 - 1] anymore: good luck for the user to understand that. Eventually, I think it would be appropriate to refactor apply_slit() and raise an error if we're applying a slit not normalized by area to a Spectrum that contains ratios (like Transmittance, but also Emissivity).  [ note that this is a different Issue ! ] 

What do you think?

pkj-m replied:
Hi @erwanp , thanks for the detailed comment. That's a lot of new information to process so I'll take it up one at a time!
Regarding why the if-else is necessary, that's a thing with astropy where concatenating strings works as follows:
```
u.Unit("1*cm")    #works fine   --> parses as u.cm
u.Unit("cm*1")    #throws error

u.Unit("mW/sr*cm")    #parses it as mW / (sr cm)
u.Unit("cm*mW/sr")    #parses it as (cm mW) / sr   --> what we want in the code
```

Regarding your second point, while I did understand the motivation and procedure behind how the experiment is conducted, I could not see how that connects with the fact that applying slit without area normalization is pointless. I did some research on the internet and the most commonly brought up argument is that peak/max normalization leads to erroneous results? Is that true?
I was also wondering if while calculating the transmittance spectra, could we perhaps allow the user to pass the intensity and background intensity separately with units? Wouldn't this help us drop "I/I0" as placeholders completely?
> but we'll end up with transmittance spectra in "nm" or "cm-1" and with values that are not within [0 - 1] anymore

Can't we normalize the spectrum after passing it through the slit?

Also apologies for any naive questions! While I did take up a few physics courses in my freshman and sophomore year, they were clearly not enough!

erwanp replied:
Regarding your first point `u.Unit("cm*1")`: it's a nice trick, but looks like a Hack! What about replacing "1" with the empty string "" ?  It seems that that's how Astropy writes u.dimensionless_unscaled units anyway: 
```
u.Unit("1").__str__()
>> ""
u.Unit("").__str__()
>> ""
u.Unit("") == u.dimensionless_unscaled
>> True
```
And no need for the Hack then! 

---

Second point:

> the most commonly brought up argument is that peak/max normalization leads to erroneous results

It's not erroneous for emission emission spectra (if you account for the change in units), and sometimes useful: if what you're interested is the maximum energy Flux, for instance in radiative transfer, you'd get a result directly in energy flux  "~W/cm2/sr", instead of spectral energy flux "~W/cm2/sr/nm" ). 
For absorption spectrum, it's erroenous as I described it in my previous comment: because the underlying quantity is a ratio, so we assume that the units of the background spectrum did not change. 

What you're describing (given intensities separately) is possible. Actually, users could do :
```
s1 = Radiance(emission_spectrum)
s2 = Radiance(background_spectrum)
transmittance = s1/s2               # I'm not sure if this work for Spectrum objects, but same is possible using s1.get(), s2.get() ? 
```
But usually your instrument, or the instrument software, will process this directly and return a transmittance spectrum directly. So the main use case takes "I/I0". 

---

>  Can't we normalize the spectrum after passing it through the slit?

We could but that's not what the user asked for, and it would become very dangerous! We better raise an error saying that applying a non-area-normalized slit to a transmittance spectrum is erroneous, and that he should do otherwise. 

We could suggest a solution in the error message: 

```
# during error handling: 
if 'radiance_noslit' in self.get_vars():
    raise ValueError("ValueError("""applying a non-area-normalized slit to a transmittance spectrum is erroneous. Use `method="area"` in apply_slit(). If you're only interested in the emission spectrum, you can extract the emission spectrum the first: `s_rad = Radiance_noslit(s).apply_slit(...)`""")
else:
    raise ValueError("ValueError("""applying a non-area-normalized slit to a transmittance spectrum is erroneous. Use `method="area"` in apply_slit()""")

```

---

Hope this is clear! 
Also whatever physics class you took does not really matter here: it's more "technicalities" (for instance, knowledge of instruments & instrument software) than physics and science at this point! :)

pkj-m replied:
Hi Erwan, thanks for the detailed reply! I understand it much better now. 
Regarding the "hack", while we could use "" instead of "1" while creating `new_unit`, but we'd still be forced to use an if-else block (albeit less complex than the current one) since even after replacing "1" with "" will result in a string such as `cm*` which when parsed by `u.Unit` returns an error. I have made this change. Let me know if you think we could do this in a better way!

For the second point, I think I can see more clearly why you'd want the user to not call the `slit` method when using `max` normalization. While I wouldn't say it's all perfectly clear, the idea and the intuition behind the problem (and it's solution) have become a lot more solid now.
Finally, the code you shared (to give an error in case the user tries to call the `apply_slit` method incorrectly) looks good to me. I think we should do that once we're done with this.

erwanp replied:
You're right about the `*`! 
Then what about this? :
```
new_unit = (u.Unit(self.units[qns]) * u.Unit(unit)).to_string()
```

Adimensioned.to_string() returns `""` so we should replace `"1"` with `""` for consistency

erwanp commented on docs/spectrum/spectrum.rst:
```diff
@@ -85,7 +85,7 @@ From numpy arrays, use :py:meth:`~radis.spectrum.spectrum.Spectrum.from_array` :
     from radis import Spectrum
     s = Spectrum.from_array(w, T, 'transmittance_noslit', 
-                               waveunit='nm', unit='I/I0')
+                               waveunit='nm', unit='1')
```
```suggestion
                               waveunit='nm', unit='')  # adimensioned
```

erwanp commented on radis/lbl/base.py:
```diff
@@ -102,9 +102,9 @@ class BaseFactory(DatabankLoader):
-        "absorbance": "-ln(I/I0)",
-        "abscoeff": "cm_1",
-        "abscoeff_continuum": "cm_1",
+        "absorbance": "1",
```
```suggestion
        "absorbance": "",
```

erwanp commented on radis/lbl/base.py:
```diff
@@ -113,10 +113,10 @@ class BaseFactory(DatabankLoader):
-        "emissivity_noslit": "eps",
-        "transmittance": "I/I0",
-        "transmittance_noslit": "I/I0",
+        "emissivity": "1",
+        "emissivity_noslit": "1",
+        "transmittance": "1",
+        "transmittance_noslit": "1",
```
```suggestion
        "emissivity": "",
        "emissivity_noslit": "",
        "transmittance": "",
        "transmittance_noslit": "",
```

erwanp commented on radis/phys/blackbody.py:
```diff
@@ -230,8 +230,8 @@ def sPlanck(
             "radiance_noslit": Iunit,
-            "transmittance_noslit": "I/I0",
-            "absorbance": "-ln(I/I0)",
+            "transmittance_noslit": "1",
+            "absorbance": "1",
```
```suggestion
            "transmittance_noslit": "",
            "absorbance": "",
```

minouHub commented:
Hi guys, 
I just read quickly the recent improvements you are implementing. I have a question regarding the unit-less quantities. It seems that @erwanp prefers `""`  instead of `"1"`. Is it technical or motivated by physics ? I looked online and it seems that `"1"` is the official way.
https://physics.stackexchange.com/questions/59978/is-there-a-symbol-for-unitless

erwanp commented:
As we're moving to Astropy system, we need to comply with their standard. They use:
```
u.dimensionless_unscaled.to_string()
>>> ""
u.Unit("1").to_string()
>>> ""
```

pkj-m commented:
@erwanp it looks like the CI server failed to procure some data for the tests due to a connection failure. Do we need to restart the build?

minouHub commented:
No no, the test just failed :)

See https://travis-ci.com/github/radis/radis/jobs/322715193

```
=========================== short test summary info ============================

FAILED radis/test/lbl/test_calc.py::test_calc_spectrum_multiple_molecules - r...

==== 1 failed, 138 passed, 7 deselected, 530 warnings in 336.66s (0:05:36) =====
```

minouHub commented:
I restarted the job at 10.21 pm :)

pkj-m commented:
Awesome! Thanks a lot. The build passed this time :)
